doc/policies/functiontype.hpp

```diff
@@ -0,0 +1,117 @@
+/*! @page function The FunctionType policy in mlpack
+
+@section Overview
+
+To represent the various types of loss functions encountered in machine
+learning problems, mlpack provides the \c FunctionType template parameter in
+the optimizer interface. The various optimizers available in the core library
+rely on this policy to gain the necessary information required by the optimizing
+algorithm.
+
+The \c FunctionType template parameter required by the Optimizer class can have
+additional requirements imposed on it, depending on the type of optimizer used.
+
+@section Interface requirements
+
+The most basic requirements for the \c FunctionType parameter are the
+implementations of two public member functions, with the following interface
+and semantics
+
+@code
+double Evaluate(const arma::mat& coordinates);
+@endcode
+
+To evaluate the loss function at the given coordinates.
+
+@code
+void Gradient(const arma::mat& coordinates, arma::mat& gradient);
+@endcode
+
+To evaluate the gradient at the given coordinates, where \c gradient is an
+out-param for the required gradient.
```

Comment:
It might be easier to read if you include this comment as part of the `@code` snippet:

```
// Evaluate the gradient at the given coordinates, where 'gradient' is an output
// parameter for the required gradient.
void Gradient(const arma::mat& coordinates, arma::mat& gradient);
```

There are lots of little `@code` blocks you could modify like this.  What do you think?